src/minimega/router.go

```diff
@@ -0,0 +1,582 @@
+	for _, d := range r.dhcp {
+		if d.low != "" {
+			fmt.Fprintf(&out, "dnsmasq dhcp range %v %v %v\n", d.addr, d.low, d.high)
+		}
+		if d.router != "" {
+			fmt.Fprintf(&out, "dnsmasq dhcp option router %v %v\n", d.addr, d.router)
+		}
+		if d.dns != "" {
+			fmt.Fprintf(&out, "dnsmasq dhcp option dns %v %v\n", d.addr, d.dns)
+		}
+		for mac, ip := range d.static {
+			fmt.Fprintf(&out, "dnsmasq dhcp static %v %v %v\n", d.addr, mac, ip)
+		}
+	}
+	for ip, host := range r.dns {
+		fmt.Fprintf(&out, "dnsmasq dns %v %v\n", ip, host)
+	}
+	for subnet, _ := range r.rad {
+		fmt.Fprintf(&out, "dnsmasq ra %v\n", subnet)
+	}
+	fmt.Fprintf(&out, "dnsmasq commit\n")
+
+	// bird
+	fmt.Fprintf(&out, "bird flush\n")
+	for network, nh := range r.staticRoutes {
+		fmt.Fprintf(&out, "bird static %v %v\n", network, nh)
+	}
+	for _, o := range r.ospfRoutes {
+		for iface, _ := range o.interfaces {
+			fmt.Fprintf(&out, "bird ospf %v %v\n", o.area, iface)
+		}
+	}
+	fmt.Fprintf(&out, "bird commit\n")
+
+	filename := filepath.Join(*f_iomBase, fmt.Sprintf("minirouter-%v", r.vmID))
+	return ioutil.WriteFile(filename, out.Bytes(), 0644)
+}
+
+// Create a new router for vm, or returns an existing router if it already
+// exists
+func FindOrCreateRouter(vm VM) *Router {
+	log.Debug("FindOrCreateRouter: %v", vm)
+
+	id := vm.GetID()
+	if r, ok := routers[id]; ok {
+		return r
+	}
+	r := &Router{
+		vmID:         id,
+		IPs:          [][]string{},
+		logLevel:     "error",
+		dhcp:         make(map[string]*dhcp),
+		dns:          make(map[string]string),
+		rad:          make(map[string]bool),
+		staticRoutes: make(map[string]string),
+		ospfRoutes:   make(map[string]*ospf),
+	}
+	nets := vm.GetNetworks()
+	for i := 0; i < len(nets); i++ {
+		r.IPs = append(r.IPs, []string{})
+	}
+
+	routers[id] = r
+
+	vm.SetTag("minirouter", fmt.Sprintf("%v", id))
+
+	return r
+}
+
+// FindRouter returns an existing router if it exists, otherwise nil
+func FindRouter(vm VM) *Router {
+	id := vm.GetID()
+	if r, ok := routers[id]; ok {
+		return r
+	}
+	return nil
+}
+
+func (r *Router) Commit() error {
+	log.Debugln("Commit")
+
+	// build a command list from the router
+	err := r.generateConfig()
+	if err != nil {
+		return err
+	}
+	r.updateIPs = false // IPs are no longer stale
+
+	// remove any previous commands
+	prefix := fmt.Sprintf("minirouter-%v", r.vmID)
+	ids := ccPrefixIDs(prefix)
+	if len(ids) != 0 {
+		for _, v := range ids {
+			c := ccNode.GetCommand(v)
+			if c == nil {
+				return fmt.Errorf("cc delete unknown command %v", v)
+			}
+
+			if !ccMatchNamespace(c) {
+				// skip without warning
+				continue
+			}
+
+			err := ccNode.DeleteCommand(v)
+			if err != nil {
+				return fmt.Errorf("cc delete command %v : %v", v, err)
+			}
+			ccUnmapPrefix(v)
+		}
+	}
+
+	// filter on the minirouter tag
+	filter := &ron.Client{
+		Tags: make(map[string]string),
+	}
+	filter.Tags["minirouter"] = fmt.Sprintf("%v", r.vmID)
+
+	// issue cc commands for this router
```

Comment:
Thinking about this more... it won't work with namespaces since it only applies to VMs on the local node. One way to solve it would be to call `cc` commands using `runCommands` which would cause a broadcast.